File: src/main/java/essentials/core/Exp.java
package essentials.core;

import mindustry.entities.type.Player;
import mindustry.gen.Call;

import static essentials.Global.config;
import static essentials.Global.nbundle;
import static essentials.core.PlayerDB.PlayerData;
import static essentials.core.PlayerDB.PlayerDataSet;

public class Exp {
    private static double BASE_XP = config.getBasexp();
    private static double EXPONENT = config.getExponent();

    public static void exp(Player player) {
        PlayerData target = PlayerData(player.uuid);

        int currentlevel = target.level;
        int max = (int) calculateFullTargetXp(currentlevel);

        int xp =  target.exp;
        int levelXp = max - xp;
        int level = calculateLevel(xp);
        int reqexp = (int)Math.floor(max);
        String reqtotalexp = xp+"("+(int) Math.floor(levelXp)+") / "+(int) Math.floor(max);

        target.reqexp = reqexp;
        target.level = level;
        target.reqtotalexp = reqtotalexp;

        if(currentlevel < level && currentlevel > config.getAlarmlevel() && config.isLevelupalarm()){
            Call.onInfoToast(nbundle("player-levelup",player.name,level),600);
        }
    }

    private static double calcXpForLevel(int level) {
        return BASE_XP+(BASE_XP * Math.pow(level, EXPONENT));
    }

    private static double calculateFullTargetXp(int level) {
        double requiredXP = 0;
        for (int i = 0; i <= level; i++) {
            requiredXP += calcXpForLevel(i);
        }
        return requiredXP;
    }

    private static int calculateLevel(double xp) {
        int level = 0;
        double maxXp = calcXpForLevel(0);
        do {
            maxXp += calcXpForLevel(++level);
        } while (maxXp < xp);
        return level;
    }
}

File: src/main/kotlin/essentials/event/feature/Exp.kt
package essentials.event.feature

import essentials.PlayerData
import essentials.data.Config
import kotlin.math.floor
import kotlin.math.pow

object Exp {
    private val baseXP = Config.baseXp
    private val exponent = Config.exponent
    private fun calcXpForLevel(level: Int): Double {
        return baseXP + baseXP * level.toDouble().pow(exponent)
    }

    private fun calculateFullTargetXp(level: Int): Double {
        var requiredXP = 0.0
        for (i in 0..level) {
            requiredXP += calcXpForLevel(i)
        }
        return requiredXP
    }

    private fun calculateLevel(xp: Double): Int {
        var level = 0
        var maxXp = calcXpForLevel(0)
        do {
            maxXp += calcXpForLevel(++level)
        } while (maxXp < xp)
        return level
    }

    operator fun get(target: PlayerData): String {
        val currentlevel = target.level
        val max = calculateFullTargetXp(currentlevel).toInt()
        val xp = target.exp
        val levelXp = max - xp
        val level = calculateLevel(xp.toDouble())
        //val reqexp = floor(max.toDouble()).toInt()
        target.level = level
        return xp.toString() + "(" + floor(levelXp.toDouble()).toInt() + ") / " + floor(max.toDouble()).toInt()
    }
}